src/test/ui/type-alias-impl-trait/self_implication.rs

```diff
@@ -0,0 +1,37 @@
+#![feature(type_alias_impl_trait)]
+fn foo() {
+    struct Foo<'a> {
+        x: &'a mut u8,
+    }
+    impl<'a> Foo<'a> {
+        fn foo(&self) -> impl Sized {}
+    }
+    // use site
+    let mut x = 5;
+    let y = Foo { x: &mut x };
+    let z = y.foo();
+    let _a = &x; // invalidate the `&'a mut`in `y`
+    let _b = z; // this should *not* check that `'a` in the type `Foo<'a>::foo::opaque` is live
+}
+
+fn bar() {
+    struct Foo<'a> {
+        x: &'a mut u8,
+    }
+
+    // desugared
+    type FooX<'a> = impl Sized;
+    impl<'a> Foo<'a> {
+        fn foo(&self) -> FooX<'a> {}
+    }
+
+    // use site
+    let mut x = 5;
+    let y = Foo { x: &mut x };
+    let z = y.foo();
+    let _a = &x; // invalidate the `&'a mut`in `y`
+    //~^ ERROR cannot borrow `x`
+    let _b = z; // this should *not* check that `'a` in the type `Foo<'a>::foo::opaque` is live
```

Comment:
that comment seems wrong? If you have `FooX<'a>` with an explicit `'a` this error feels correct? :thinking: 

can you add a comment to the top of the file explaining what we're testing here?

Reply:
> that comment seems wrong? If you have `FooX<'a>` with an explicit `'a` the error feels correct?

that's the thing... lifetimes on type aliases should not matter for anything. Only lifetimes in the aliased type. Once/if we avoid adding uncaptured lifetimes to opaque types, this would automatically end up compiling. It may end up working with https://github.com/rust-lang/rust/pull/103491 or with that plus some small changes.